### PR TITLE
Update and expand information about building Gitea

### PR DESCRIPTION
**Pull request by zeripath**
The current information on gitea.io for building and hacking on is inadequate. This PR expands on this and provides some more information.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 1]
```diff
@@ -15,41 +15,258 @@ menu:
...
+You should [install go](https://golang.org/doc/install) and set up your go
+environment correctly. In particular, it is recommended to set the `$GOPATH`
+environment variable and to add the go bin directory or directories
+`${GOPATH//://bin:}/bin` to the `$PATH`. See the Go wiki entry for
+[GOPATH](https://github.com/golang/go/wiki/GOPATH).
 
-Some internal packages are referenced using their respective Github URL. This can
-become problematic. To "trick" the Go tool into thinking this is a clone from the
-official repository, download the source code using "go get" and follow these instructions.
+**Note**: Whilst building, Gitea will download and build several packages
```
Comment (kolaente) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md]: Isn't this true only for Go 1.11 or stuff like `make lint`?

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 2]
```diff
@@ -15,41 +15,258 @@ menu:
...
 ```
 
-Fork the [Gitea repository](https://github.com/go-gitea/gitea) on GitHub, it should
-then be possible to switch the source directory on the command line.
+This will clone the Gitea source code to: `"$GOPATH"/src/code.gitea.io/gitea`, or if `$GOPATH`
+is not set `"$HOME"/go/src/code.gitea.io/gitea`.
 
+## Forking Gitea
+
+As stated above, you cannot clone Gitea to an arbitrary path. Download the master Gitea source
```
Comment (kolaente) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md:67]: Maybe add that this is't the case with Go 1.11? Or should we wait until we fully support Go modules?

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 3]
```diff
@@ -15,41 +15,258 @@ menu:
...
+git fetch --all --prune
 ```
 
 To be able to create pull requests, the forked repository should be added as a remote
 to the Gitea sources, otherwise changes can't be pushed.
 
+## Building Gitea (Basic)
+
+Take a look at our
+<a href='{{< relref "doc/installation/from-source.en-us.md" >}}'>instructions</a>
```
Comment (kolaente) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md:94]: Can this be a markdown link?

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 4]
```diff
@@ -15,41 +15,258 @@ menu:
...
+for <a href='{{< relref "doc/installation/from-source.en-us.md" >}}'>building
```
Comment (kolaente) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md:95]: Can this be a markdown link?

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 5]
```diff
@@ -15,41 +15,258 @@ menu:
...
+
+### Updating the API
+
+When creating new API routes or modifying existing API routes you **MUST**
+update and/or create [Swagger](https://swagger.io/docs/specification/2-0/what-is-swagger/)
+documentation for these. You should be careful not to break the API for
+downstream users which depend on a stable API. In general this means additions
+are acceptable, but deletions or fundamental changes of API will be rejected.
+
+Once you have created or changed an API endpoint, please regenerate the Swagger
```
Comment (kolaente) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md:176]: We should also add a link to the library we use to genereate the docs from comments and about the comment syntax.

**Review by kolaente — COMMENTED** [review]
Looks good so far, some small nits though.

Could you also add a note about adding new config options to the docs? (mainly where to put them)

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 1 above)
Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md]: so `make lint` is actually a thin wrapper around `revive` and `go` will get the tool as necessary. 

Or did you mean something else? I guess I've not made it clear that you will need to have make installed?

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 2 above)
Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md:67]: When we move to go modules fully we can update this - but for the moment I think we should leave it at this.

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 3 above)
Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md:94]: AFAIK nope.

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 4 above)
Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md:95]: Similarly.

File: docs/content/doc/installation/from-source.en-us.md [hunk 6]
```diff
@@ -15,68 +15,95 @@ menu:
...
+following commands to fetch the source and switch into the source directory.
+Go is quite opinionated about where it expects its source code, and simply
+cloning the Gitea repository to an arbitrary path is likely to lead to
+problems - the fixing of which is out of scope for this document.
 
-```
+```bash
 go get -d -u code.gitea.io/gitea
-cd $GOPATH/src/code.gitea.io/gitea
+cd "$GOPATH"/src/code.gitea.io/gitea
```
Comment (lafriks) [inline docs/content/doc/installation/from-source.en-us.md]: I think ending quote should be at the and of line

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 5 above)
Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md:176]: Do you mean more swagger documentation or something else?

File: docs/content/doc/installation/from-source.en-us.md (hunk 6 above)
Comment (zeripath) [reply docs/content/doc/installation/from-source.en-us.md]: It doesn't really matter AFAIU, but I think keeping the quote next to variable keeps it with its context. 

If we had `"$GOPATH/src/code gitea io/gitea"` then we would have to extend the quoted section, but as it stands the quote just needs to be around bit that could possibly have a space. I would probably prefer that to be `"$GOPATH"/src/"code gite io"/gitea` though.

If you much prefer extending the quote then I will change.

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 7]
```diff
@@ -15,41 +15,258 @@ menu:
...
+cloning the Gitea repository to an arbitrary path is likely to lead to
+problems - the fixing of which is out of scope for this document. Further some
+internal packages are referenced using their respective Github URL and at
+present we use `vendor/` directories.
+
+The recommended method of obtaining the source code is by using the `go get` command:
+
+```bash
 go get -d code.gitea.io/gitea
+cd "$GOPATH"/src/code.gitea.io/gitea
```
Comment (sapk) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md]: ~Maybe use $(go env GOPATH) ? As it will also work when GOPATH is not set.~
Ignore as we request to set $GOPATH at start.

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 5 above)
Comment (sapk) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md:176]: Maybe link to https://goswagger.io/use/spec.html#annotation-syntax

Comment (sapk) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md:176]: Or a PR that add an endpoint: https://github.com/go-gitea/gitea/pull/5843/files

**Comment by zeripath** [comment]
I've made some changes to hacking:

I've added information about adding settings. (Albeit perhaps a little cursory as I note that @lunny has a current PR open to refactor `modules/setting/setting.go` )

I've added a note to say you will need make (I thought I had put that in already. oops)

I've put a bit more detail in about swagger and go-swagger.

File: docs/content/doc/installation/from-source.en-us.md (hunk 6 above)
Comment (adelowo) [reply docs/content/doc/installation/from-source.en-us.md]: is the quote even needed?

Comment (zeripath) [reply docs/content/doc/installation/from-source.en-us.md]: Yeah if $GOPATH contains a space. It's good practice to put environment variables in quotes in general.

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 1 above)
Comment (kolaente) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md]: I don't mean make, I think this implies when you run `go build` (either directly or through the makefile, Gitea will download all go dependencies automatically. This is not the case, only for things like `make lint`/`make misspell-check`/etc.

Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md]: What wording would you suggest?

Comment (kolaente) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md]: Something like `when executing tasks which require external tools, like make misspell-check, Gitea will automatically download and build these tools. To be able for them to work properly, you need to add the go bin (usually $GOPATH/bin) to your path.`

And we should probably put a note somewhere about the facte that we're keeping all dependencies in the `vendor` folder while we're at it.

Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md]: Hmm. I don't think I know enough about `dep ensure` and the like to go there. Perhaps someone could do a follow up PR or suggest a section to put in wholesale about this.

**Comment by zeripath** [comment]
Adjusted the clarifying note about automatic build dependencies.

I have avoided mentioning vendor so far because I expect that could be quite complex. We would likely need to talk about dep ensure, make test-vendor and gopkg.toml.

There is in general a lack of documentation about hacking on the UI too. This could be done as another pr.

File: docs/content/doc/installation/from-source.en-us.md (hunk 6 above)
Comment (lafriks) [reply docs/content/doc/installation/from-source.en-us.md]: I think all path must be quoted if any part of it contains space "$GOPATH/src/code gite io/gitea" so that it is parsed as single command argument

**Comment by sapk** [comment]
Vendoring and GOPATH will also change when we migrate to `go mod`

**Review by sapk — APPROVED** [review]

File: docs/content/doc/installation/from-source.en-us.md (hunk 6 above)
Comment (zeripath) [reply docs/content/doc/installation/from-source.en-us.md]:
```bash
export TD="A Test Directory"
mkdir "$TD"
echo "foo" > "$TD"/foo
mkdir -p "$TD"/no-space/"Some Spaces"
```

Comment (lafriks) [reply docs/content/doc/installation/from-source.en-us.md]: personally have not used or seen anywhere used such syntax... putting quotes around full path seems more natural to me but that is just my personal preference

Comment (zeripath) [reply docs/content/doc/installation/from-source.en-us.md]: OK I've extended the quotes.

Comment (lafriks) [reply docs/content/doc/installation/from-source.en-us.md]: thanks :)

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md [hunk 8]
```diff
@@ -15,41 +15,280 @@ menu:
...
+is recommended that you allow Gitea Managers and Owners to modify your PR
+branches as we will need to update it to master before merging and/or may be
+able to help fix issues directly.
+
+Any PR requires two approvals from the Gitea maintainers and needs to pass the
+continous integration. Take a look at our
+[`CONTRIBUTING.md`](https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md)
+document.
+
+If you need more help pop on to [Discord](https://discord.gg/NsatcWJ) #Develop
```
Comment (techknowlogick) [inline docs/content/doc/advanced/hacking-on-gitea.en-us.md]: Can you update this to vanity Gitea discord URL (discord.gg/gitea)?

File: docs/content/doc/installation/from-source.en-us.md [hunk 9]
```diff
@@ -15,68 +15,97 @@ menu:
...
+`${GOPATH//://bin:}/bin` to the `$PATH`. See the Go wiki entry for
+[GOPATH](https://github.com/golang/go/wiki/GOPATH).
+
+**Note**: When executing make tasks that require external tools, like
+`make misspell-check`, Gitea will automatically download and build these as
+necessary. To be able to use these you must have the `"$GOPATH/bin"` directory
+on the executable path. If you don't add the go bin directory to the
+executable path you will have to manage this yourself.
+
+**Note 2**: Go version 1.8 or higher is required, however it is recommended to
```
Comment (techknowlogick) [inline docs/content/doc/installation/from-source.en-us.md]: ah darn, when we deprecated 1.8 it seems we missed this in documentation. Can you set this to 1.9?

Although now that 1.12 is coming out in a week or two we should probably look at deprecating 1.9 too (but that is out of scope of this PR)

**Review by techknowlogick — COMMENTED** [review]
Thanks for updating the docs!! 🎊 

A few nits, but otherwise looks fantastic!

File: docs/content/doc/advanced/hacking-on-gitea.en-us.md (hunk 8 above)
Comment (zeripath) [reply docs/content/doc/advanced/hacking-on-gitea.en-us.md]: Done.

File: docs/content/doc/installation/from-source.en-us.md (hunk 9 above)
Comment (zeripath) [reply docs/content/doc/installation/from-source.en-us.md]: Done.

**Review by techknowlogick — APPROVED** [review]

**Comment by zeripath** [comment]
I just noticed that we'd added postcss so added a little thing about that. I think it's probably ready to merge now.

**Review by kolaente — APPROVED** [review]

**Comment by zeripath** [comment]
Thanks peeps. Hopefully people will find this helpful. I guess we just need to try to keep it up to date.

I'll pop a comment on @Typeless go module pr to suggest that a bit of information to hacking and install is put down.
